ledger/participant-state-index/reference/src/main/scala/com/daml/ledger/participant/state/index/v1/impl/reference/ReferenceIndexService.scala

```diff
@@ -86,17 +87,19 @@ final case class ReferenceIndexService(
       .to(updateStateSink)
       .run()
 
-  private def asyncResultWithState[T](ledgerId: LedgerId)(
-      handler: IndexState => Future[T]): AsyncResult[T] = {
+  // Return a result using a given handler function
+  // Waits for state to be initialized (e.g. first
+  // update to arrive in order to established ledger beginning and current end) before calling the handler.
+  private def futureWithState[T](handler: IndexState => Future[T]): Future[T] = {
```

Comment:
Nitpick: The fact that this function is necessary to be used in so many places feels wrong to me. It means that there's too much dependence on the mutable state here. This also means that all of these calls are completely async and have unpredictable completion time. Few ideas to explore:
- make it acceptable for the interface of this service to return "not ready" error when called;
- split the implementation into two classes: one for uninitialized state, which always returns "not ready" errors and one for initialized state which returns what it should.

I think this will make the implementation cleaner (but may make the interface harder to use). Devil is in the details. Up to you.

Reply:
They are inherently asynchronous as the proper implementation likely will involve SQL queries.

Reply:
IMO, the async approach on the service side is better. An interface serving "not ready" error pushes clients to implementing an async layer on their side which leads to code repetition and errors. So I would not change anything.